codersdk/metrics.go

```diff
@@ -0,0 +1,145 @@
+package codersdk
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+	"net/http/cookiejar"
+	"time"
+
+	"golang.org/x/xerrors"
+	"nhooyr.io/websocket"
+	"nhooyr.io/websocket/wsjson"
+
+	"cdr.dev/slog"
+	"github.com/coder/coder/agent"
+	"github.com/coder/retry"
+)
+
+type CloseFunc func() error
+
+func (c CloseFunc) Close() error {
+	return c()
+}
+
+// AgentReportStats begins a stat streaming connection with the Coder server.
+// It is resilient to network failures and intermittent coderd issues.
+func (c *Client) AgentReportStats(
+	ctx context.Context,
+	log slog.Logger,
+	stats func() *agent.Stats,
+) (io.Closer, error) {
+	serverURL, err := c.URL.Parse("/api/v2/metrics/report-agent-stats")
+	if err != nil {
+		return nil, xerrors.Errorf("parse url: %w", err)
+	}
+
+	jar, err := cookiejar.New(nil)
+	if err != nil {
+		return nil, xerrors.Errorf("create cookie jar: %w", err)
+	}
+
+	jar.SetCookies(serverURL, []*http.Cookie{{
+		Name:  SessionTokenKey,
+		Value: c.SessionToken,
+	}})
+
+	httpClient := &http.Client{
+		Jar: jar,
+	}
+
+	doneCh := make(chan struct{})
+	ctx, cancel := context.WithCancel(ctx)
+
+	go func() {
+		defer close(doneCh)
+
+		for r := retry.New(time.Second, time.Hour); r.Wait(ctx); {
```

Comment:
Does this retry period reset? Or once it grows to 1 hour, it'll always be an hour?

Like if coderd goes down, the growth grows to 1 hr, then coderd comes back. Will it wait for 1 hour the next time coderd goes down?

If so, we should manually reset `r` when we succeed.

Reply:
Good point